BeatTogether.DedicatedServer.Kernel/Managers/GameplayManager.cs

```diff
@@ -124,46 +127,49 @@ public async void StartSong(BeatmapIdentifier beatmap, GameplayModifiers modifie
             songReadyCts.CancelAfter((int)(SongLoadTimeLimit * 1000));
             await Task.WhenAll(songReadyTasks);
 
-            // If no players are actually playing
-            if (_playerRegistry.Players.All(player => !player.InGameplay))
-            {
-                _requestReturnToMenuCts.Cancel();
-            }
-
+            // If no players are actually playing, or not all players are not in the lobby(if at least one player is then true)
+            if (loadingPlayers.All(player => !player.InGameplay) || !loadingPlayers.All(player => !player.InLobby))
+                _requestReturnToMenuCts.Cancel(); //this will cancel the gameplay if someone is in the lobby
+            
             // Start song and wait for finish
             State = GameplayManagerState.Gameplay;
             _songStartTime = _instance.RunTime + SongStartDelay;
             _packetDispatcher.SendToNearbyPlayers(new SetSongStartTimePacket
             {
                 StartTime = _songStartTime
             }, DeliveryMethod.ReliableOrdered);
-            await Task.WhenAll(levelFinishedTasks);
 
+            await Task.WhenAll(levelFinishedTasks);
             State = GameplayManagerState.Results;
 
             // Wait at results screen if anyone cleared
             if (_levelCompletionResults.Values.Any(result => result.LevelEndStateType == LevelEndStateType.Cleared))
                 await Task.Delay((int)(ResultsScreenTime * 1000));
 
             // End gameplay and reset
-            _playerSpecificSettings.Clear();
-            _levelCompletionResults.Clear();
+            ResetValues(null, null);
+            State = GameplayManagerState.None;
+            _packetDispatcher.SendToNearbyPlayers(new ReturnToMenuPacket(), DeliveryMethod.ReliableOrdered); //quest ignores this, has already returned to the lobby
+            _instance.SetState(MultiplayerGameState.Lobby);
+        }
+
+        private void ResetValues(BeatmapIdentifier? map, GameplayModifiers? modifiers)

```

Comment:
Reset that is called before and after the gameplay is run. Clears/sets the beatmap and modifiers and resets tasks. Runs at beginning so _levelCompletionResults.Clear(); is run before as it was not being cleared if the gameplay ended early (if everyone failed it seemed to not be ran, so if the player tried playing again then they would instantly fail). By running this at the beginning too it clears the completionresults before the level is ran.

Reply:
BUGFIX